src/compiler/checker.ts

```diff
@@ -26107,6 +26107,9 @@ namespace ts {
                     checkTypeComparableTo(exprType, targetType, errNode,
                         Diagnostics.Conversion_of_type_0_to_type_1_may_be_a_mistake_because_neither_type_sufficiently_overlaps_with_the_other_If_this_was_intentional_convert_the_expression_to_unknown_first);
                 }
+                else if (targetType.flags & TypeFlags.StringLiteral && isStringLiteral(expression) && expression.text !== (targetType as StringLiteralType).value) {

```

Comment:
few questions / suggestions:
* why limit it to assertions on a string literal expression?
  * should probably apply to all literal types
  * numeric literal
  * boolean literals
  * bigint literals
* check union types:
  * `'foo' as 'foo' | 'bar'` -> Ok
  * `'foo' as 'bar' | 'baz'` -> Error
  * `'foo' | 'bar' as 'foo'` -> Ok
  * `'bar' | 'baz' as 'foo'` -> Error
  * `'foo' | 'bar' as 'foo' | 'baz'` -> Ok
  * `'foo' | 'bar' as 'baz' | 'bas'` -> Error

Reply:
```suggestion
                else if (isIntersectionEmpty(targetType, exprType)) {

```

?

Reply:
It's just a proof of concept right now. Also, no, not that because the `exprType` gets widened before this point which is the main issue.

Reply:
If they were widened, then they wouldn't be literals anyway anymore, no?